packages/components/src/number-control/index.tsx

```diff
@@ -0,0 +1,209 @@
+		( state, action ) => {
+			const nextState = { ...state };
+
+			const { type, payload } = action;
+			const event = payload.event;
+			const currentValue = nextState.value;
+
+			/**
+			 * Handles custom UP and DOWN Keyboard events
+			 */
+			if (
+				type === inputControlActionTypes.PRESS_UP ||
+				type === inputControlActionTypes.PRESS_DOWN
+			) {
+				const enableShift =
+					( event as KeyboardEvent | undefined )?.shiftKey &&
+					isShiftStepEnabled;
+
+				const incrementalValue = enableShift
+					? ensureNumber( shiftStep ) * baseStep
+					: baseStep;
+				let nextValue = isValueEmpty( currentValue )
+					? baseValue
+					: currentValue;
+
+				if ( event?.preventDefault ) {
+					event.preventDefault();
+				}
+
+				if ( type === inputControlActionTypes.PRESS_UP ) {
+					// @ts-expect-error TODO: Investigate if this is wrong
+					nextValue = add( nextValue, incrementalValue );
+				}
+
+				if ( type === inputControlActionTypes.PRESS_DOWN ) {
+					// @ts-expect-error TODO: Investigate if this is wrong
+					nextValue = subtract( nextValue, incrementalValue );
+				}
+
+				// @ts-expect-error TODO: Investigate if this is wrong
+				nextState.value = constrainValue(
+					// @ts-expect-error TODO: Investigate if this is wrong
+					nextValue,
+					enableShift ? incrementalValue : undefined
+				);
+			}
+
+			/**
+			 * Handles drag to update events
+			 */
+			if ( type === inputControlActionTypes.DRAG && isDragEnabled ) {
+				// @ts-expect-error TODO: Investigate
+				const [ x, y ] = payload.delta;
+				// @ts-expect-error TODO: Investigate
+				const enableShift = payload.shiftKey && isShiftStepEnabled;
+				const modifier = enableShift
+					? ensureNumber( shiftStep ) * baseStep
+					: baseStep;
+
+				let directionModifier;
+				let delta;
+
+				switch ( dragDirection ) {
+					case 'n':
+						delta = y;
+						directionModifier = -1;
+						break;
+
+					case 'e':
+						delta = x;
+						directionModifier = isRTL() ? -1 : 1;
+						break;
+
+					case 's':
+						delta = y;
+						directionModifier = 1;
+						break;
+
+					case 'w':
+						delta = x;
+						directionModifier = isRTL() ? 1 : -1;
+						break;
+				}
+
+				if ( delta !== 0 ) {
+					delta = Math.ceil( Math.abs( delta ) ) * Math.sign( delta );
+					const distance = delta * modifier * directionModifier;
+
+					// @ts-expect-error TODO: Investigate if this is wrong
+					nextState.value = constrainValue(
+						// @ts-expect-error TODO: Investigate if this is wrong
+						add( currentValue, distance ),
+						enableShift ? modifier : undefined
+					);
+				}
+			}
+
+			/**
+			 * Handles commit (ENTER key press or blur)
+			 */
+			if (
+				type === inputControlActionTypes.PRESS_ENTER ||
+				type === inputControlActionTypes.COMMIT
+			) {
+				const applyEmptyValue =
+					required === false && currentValue === '';
+
+				// @ts-expect-error TODO: Investigate if this is wrong
```

Comment:
Instead of "Investigate if this is wrong", should the comment be more about the fact that this error will only be solved once we make a decision on the `value` type?

(same for the rest of the `@ts-expect-error`  comments about the type of the `value` prop)

Reply:
Good call, I improved the TODO comments all around.

One of the interesting causes was the [`isValueEmpty()`](https://github.com/WordPress/gutenberg/blob/90948f0b1c9c1cace590e39d436336c74ffb57ed/packages/components/src/utils/values.js#L24) util function. I was able to fix the typing, but it's a bit out of scope for this PR so I'll submit it separately.